src/compiler/core.ts

```diff
@@ -1399,27 +1399,30 @@ namespace ts {
     /** Returns lower case string */
     export function toLowerCase(x: string) { return x.toLowerCase(); }
 
-    const fileNameLowerCaseRegExp = /[^İi̇ıßa-z0-9\\/:\-_\. ]+/g;
+    // We convert the file names to lower case as key for file name on case insensitive file system
+    // While doing so we need to handle special characters (eg \u0130) to ensure that we dont convert
+    // it to lower case, fileName with its lowercase form can exist along side it.

```

Comment:
Is this windows specific (does this need to live in `sys` and keyed off of `process.os`), or is this also important on mac?

Reply:
This is a feature of the standard Windows filesystem (NTFS). It's not affected by system locale - the upcase table is hardcoded into the filesystem. In theory, other case-insensitive filesystems could behave differently re: casing. 

I'm a lot less clear on the Mac OS system specifics but I just tried out creating some files on my Macbook, and a similar issue exists - (I,i) , (İ) and (ı) are all distinct filenames even though the filesystem is case insensitive. Note this is slightly different than Windows which treats İ and ı as equivalent.

Reply:
Is the behavior on Windows different if the active drive is backed by FAT32 instead of NTFS (eg, an old thumb drive)? Or do FAT and NTFS have the same filename case handling in windows?

Reply:
I'm like 99.99% sure FAT32 behaves exactly the same but I'm speaking from memory.

Reply:
This is what it looks like on node for Mac:

```sh
❯ node
> "İi̇ı".toLowerCase()
'i̇i̇ı'
> "İi̇ı".toUpperCase()
'İİI'
```

Reply:
We're interested in filesystem behavior (what's equivalent to what on disk) - `toLowerCase` and `toUpperCase` in the language are specified :P

Reply:
Thanks orta for looking into that.
I don't think, we should depend on the os to determine how to get canonical file name at all.. Considering `İı` as unique(that is not changing file casing for those) is right thing to do.. In worst cases we wont recognize that file with `İ` and `i̇` are not same which would be really a small group. The `GetCanonicalFileName` is internal compiler thing that currently is based on if file system is case sensitive or not and we shouldn't change that fact. It is an ok trade off to do in my opinion

cc: @RyanCavanaugh  and @DanielRosenwasser to see if they agree.

Reply:
Also we do not want to use `toUpperCase` since in most cases file name is lowercase so not having to store duplicate strings just for case swapping is more performant

Reply:
> I don't think, we should depend on the os to determine how to get canonical file name at all

No, I'm not saying query the OS - i'm saying check `process.os` (in `sys.ts`, as that's where we do things like that) to swap between a windows implementation of `toFileNameLowerCase` and a mac implementation of `toFileNameLowerCase`, since their handling of turkish i differs. As-is, this might fix turkish i on windows, but will potentially break it on mac, since our existing `toLowerCase` call _happened_ to already to the right thing for the mac's implementation of case-insensitivity.

Pretty much: we shouldn't break one platform to fix another - we gotta handle both.

Reply:
The argument is pretty much that our discussion here has made it evident that, as it relates to filesystems, there isn't just `case sensitive` vs `case insensitive` - there's `case sensitive` vs `windows case insensitive` vs `mac case insensitive` (probably vs `linux bespoke filesystem case insensitive` but we care about that way, way less), and while the later two are they same in the ASCII range, they differ in some of the edge cases, as @minestarks pointed out. We should strive to handle them both.